src/EnergyPlus/FanCoilUnits.cc

```diff
@@ -762,6 +763,50 @@ namespace FanCoilUnits {
                     ShowContinueError(state, "..Fan coil unit air outlet node name = " + state.dataLoopNodes->NodeID(FanCoil(FanCoilNum).AirOutNode));
                     ErrorsFound = true;
                 }
+                bool ZoneNodeNotFound = true;
+                for (CtrlZone = 1; CtrlZone <= state.dataGlobal->NumOfZones; ++CtrlZone) {
+                    if (!state.dataZoneEquip->ZoneEquipConfig(CtrlZone).IsControlled) continue;
+                    for (NodeNum = 1; NodeNum <= state.dataZoneEquip->ZoneEquipConfig(CtrlZone).NumExhaustNodes; ++NodeNum) {
+                        if (FanCoil(FanCoilNum).AirInNode == state.dataZoneEquip->ZoneEquipConfig(CtrlZone).ExhaustNode(NodeNum)) {
+                            ZoneNodeNotFound = false;
+                            break;
+                        }
+                    }
+                    if (!ZoneNodeNotFound) break;
+                }
+                if (ZoneNodeNotFound) {
+                    // Varify zone number and return air node
+                    for (CtrlZone = 1; CtrlZone <= state.dataGlobal->NumOfZones; ++CtrlZone) {
+                        for (int Num = 1; Num <= state.dataZoneEquip->ZoneEquipList(CtrlZone).NumOfEquipTypes; ++Num) {
+                            if (UtilityRoutines::SameString(FanCoil(FanCoilNum).Name, state.dataZoneEquip->ZoneEquipList(CtrlZone).EquipName(Num)) &&
+                                UtilityRoutines::SameString(FanCoil(FanCoilNum).UnitType,
+                                                            state.dataZoneEquip->ZoneEquipList(CtrlZone).EquipType(Num))) {
+                                FanCoil(FanCoilNum).ControlZoneNum = CtrlZone;
+                                break;
+                            }
+                        }
+                        if (FanCoil(FanCoilNum).ControlZoneNum > 0) break;
```

Comment:
If this search to set `FanCoil(FanCoilNum).ControlZoneNum` was done before all of the node checking, then the loops for `CtrlZone` could be eliminated.
Also, the inlet node checks are the same for ATMixer_SupplySide or no mixer, so this can be structured to avoid duplicate code.

Reply:
@mjwitte I debug the code and found FanCoil(FanCoilNum).ControlZoneNum is not assigned. There is a section to get ControlZoneNum without a mixer. Therefore, this section code is needed.

I agree that the inlet node checks are duplicated. I am going to created a new function.

Reply:
It would be better to always set `ControlZoneNum` before all of this connection checking. I don't think that would break anything else.

Reply:
Yes, it can be done. However, this search is needed with ATMixer_SupplySide and an induced node is used as an inlet node. Therefore, this section is better to stay as is.

Reply:
My point here is that I see three new loops over `CtrlZone`. It seems that one loop would be adequate.